crates/hotshot/src/lib.rs

```diff
@@ -213,16 +213,17 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> SystemContext<TYPES, I> {
             saved_leaves.insert(leaf.commit(), leaf.clone());
         }
         if let Some(payload) = anchored_leaf.get_block_payload() {
-            let encoded_txns: Vec<u8> = match payload.encode() {
-                // TODO (Keyao) [VALIDATED_STATE] - Avoid collect/copy on the encoded transaction bytes.
-                // <https://github.com/EspressoSystems/HotShot/issues/2115>
-                Ok(encoded) => encoded.into_iter().collect(),
+            let encoded_txns = match payload.encode() {
+                Ok(encoded) => encoded,
                 Err(e) => {
                     return Err(HotShotError::BlockError { source: e });
                 }
             };
 
-            saved_payloads.insert(anchored_leaf.get_view_number(), encoded_txns);
+            saved_payloads.insert(
+                anchored_leaf.get_view_number(),
+                encoded_txns.as_ref().to_vec(),
```

Comment:
omg I wasn't even aware that we're copying the payload here anyway. Oh well, it's good practice to clone it only when it's actually needed